spam/03_spam_data_slicing_tutorial.py

```diff
@@ -362,80 +249,198 @@ def textblob_polarity(x):
 #
 # A technique like upsampling might work with a small number of slices, but with hundreds or thousands or production slices, it could quickly become intractable to tune upsampling weights per slice.
 # In the following section, we show a modeling approach that we call _Slice-based Learning,_ which handles numerous slices using with slice-specific representation learning.
 
+# %% [markdown]
+# ### Set up modeling pipeline with `BinarySlicingClassifier`
+#
+# Snorkel supports performance monitoring on slices using discriminative models from `snorkel.slicing`.
+# To demonstrate this functionality, we'll first set up a the datasets + modeling pipeline in the PyTorch-based [`snorkel.classification`](https://snorkel.readthedocs.io/en/redux/packages/classification.html) package.
+#
+# *For more about this framework, please see the [`Multitask`](https://github.com/snorkel-team/snorkel-tutorials/blob/master/multitask/multitask_tutorial.ipynb) tutorial.*
+
+# %% [markdown]
+# First, we initialize a [`DictDataLoader`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.DictDataLoader.html) for each split.
+# This class in [`snorkel.classification`](https://snorkel.readthedocs.io/en/redux/packages/classification.html) that provides extra flexibility with a dictionaries of data fields (`X_dict`) and labels (`Y_dict`).
+
+# %%
+from utils import create_dict_dataloader
+
+BATCH_SIZE = 64
+
+
+train_dl = create_dict_dataloader(
+    X_train, Y_train, "train", batch_size=BATCH_SIZE, shuffle=True
+)
+valid_dl = create_dict_dataloader(
+    X_valid, Y_valid, "valid", batch_size=BATCH_SIZE, shuffle=False
+)
+test_dl = create_dict_dataloader(
+    X_test, Y_test, "test", batch_size=BATCH_SIZE, shuffle=True
+)
+
+# %% [markdown]
+# We'll inspect some of the dictionary keys below:
+
+# %%
+train_dl.dataset
+
+# %% [markdown]
+# We'll now initialize a `BinarySlicingClassifier`:
+# * `base_architecture`: We define a simple Multi-Layer Perceptron (MLP) in Pytorch to serve as the primary representation architecture. We note that the `BinarySlicingClassifier` is **agnostic to the base architecture** — you might leverage a Transformer model for text, or a ResNet for images.
+# * `head_dim`: identifies the final output feature dimension of the `base_architecture`
+# * `input_data_key`: corresponds to the desired input field from the `X_dict`
+# * `task_name`: corresponds to the corresponding input field from the `Y_dict`
+
+# %%
+from snorkel.slicing import BinarySlicingClassifier
+from utils import get_pytorch_mlp
+
+
+# Define model architecture
+bow_dim = X_train.shape[1]
+mlp = get_pytorch_mlp(hidden_dim=bow_dim, num_layers=2)
+
+# Init slice model
+slice_model = BinarySlicingClassifier(
+    base_architecture=mlp,
+    head_dim=bow_dim,
+    input_data_key="bow_features",
+    task_name="spam_task",
+    slice_names=[sf.name for sf in sfs],
+)
```

Comment:
What are the arguments against simplifying this further by dropping the need to input `head_dim`, `input_data_key`, and `task_name`?

Everyone on the (ML) planet is used to `X` being a `n x d` where `d` is the feature dim- so will be confused by the first two / is just overkill as a default- and ditto with task_name...

Reply:
Also: where the heck did `S` go???  How is it not here?

Reply:
Okay reading ahead I get that you are (1) first training a normal classifier and just monitoring slice performance during training, and (2) then training a slice-aware model.  Skip (1)- it's a needless and potentially confusing detail, since (2) includes both